sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java

```diff
@@ -0,0 +1,878 @@
+
+      public int getNumShards() {
+        return numShards;
+      }
+
+      public String getFilenameSuffix() {
+        return filenameSuffix;
+      }
+
+      public boolean needsValidation() {
+        return validate;
+      }
+    }
+  }
+
+  /**
+   * Possible TFRecord file compression types.
+   */
+  public enum CompressionType {
+    /**
+     * Automatically determine the compression type based on filename extension.
+     */
+    AUTO(""),
+    /**
+     * Uncompressed.
+     */
+    NONE(""),
+    /**
+     * GZipped.
+     */
+    GZIP(".gz"),
+    /**
+     * Deflate compressed.
+     */
+    DEFLATE(".deflate");
+
+    private String filenameSuffix;
+
+    CompressionType(String suffix) {
+      this.filenameSuffix = suffix;
+    }
+
+    /**
+     * Determine if a given filename matches a compression type based on its extension.
+     * @param filename the filename to match
+     * @return true iff the filename ends with the compression type's known extension.
+     */
+    public boolean matches(String filename) {
+      return filename.toLowerCase().endsWith(filenameSuffix.toLowerCase());
+    }
+  }
+
+  // Pattern which matches old-style shard output patterns, which are now
+  // disallowed.
+  private static final Pattern SHARD_OUTPUT_PATTERN = Pattern.compile("@([0-9]+|\\*)");
+
+  private static void validateOutputComponent(String partialFilePattern) {
+    checkArgument(
+        !SHARD_OUTPUT_PATTERN.matcher(partialFilePattern).find(),
+        "Output name components are not allowed to contain @* or @N patterns: "
+            + partialFilePattern);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  /** Disable construction of utility class. */
+  private TFRecordIO() {}
+
+  /**
+   * A {@link FileBasedSource} which can decode records in TFRecord files.
+   */
+  @VisibleForTesting
+  static class TFRecordSource extends FileBasedSource<byte[]> {
+    @VisibleForTesting
+    TFRecordSource(String fileSpec) {
+      super(fileSpec, 1L);
+    }
+
+    @VisibleForTesting
+    TFRecordSource(ValueProvider<String> fileSpec) {
+      super(fileSpec, Long.MAX_VALUE);
+    }
+
+    private TFRecordSource(String fileName, long start, long end) {
+      super(fileName, Long.MAX_VALUE, start, end);
+    }
+
+    @Override
+    protected FileBasedSource<byte[]> createForSubrangeOfFile(
+        String fileName,
+        long start,
+        long end) {
+      checkArgument(start == 0, "TFRecordSource is not splittable");
+      return new TFRecordSource(fileName, start, end);
+    }
+
+    @Override
+    protected FileBasedReader<byte[]> createSingleFileReader(PipelineOptions options) {
+      return new TFRecordReader(this);
+    }
+
+    @Override
+    public Coder<byte[]> getDefaultOutputCoder() {
+      return DEFAULT_BYTE_ARRAY_CODER;
+    }
+
+    @Override
+    protected boolean isSplittable() throws Exception {
+      // TFRecord files are not splittable
+      return false;
```

Comment:
Do you have a link to the splittable spec?
I followed this doc and didn't see anything about a marker. https://www.tensorflow.org/api_guides/python/python_io#tfrecords_format_details